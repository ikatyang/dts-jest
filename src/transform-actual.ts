import { config_namespace, AssertionFlag, JestConfig } from './definitions';
import { create_expecteds } from './utils/create-expecteds';
import { create_triggers } from './utils/create-triggers';
import { get_config } from './utils/get-config';
import { get_formatted_description } from './utils/get-formatted-description';
import { rewrite_expecteds_method } from './utils/rewrite-expecteds-method';

export const transform_actual: jest.Transformer['process'] = (
  source_text,
  source_filename,
  jest_config: JestConfig,
) => {
  const { typescript: ts, tsconfig } = get_config(
    jest_config.globals[config_namespace],
    jest_config.rootDir,
  );

  const source_file = ts.createSourceFile(
    source_filename,
    source_text,
    ts.ScriptTarget.Latest,
    false,
  );
  const triggers = create_triggers(source_file, ts);

  const expecteds = create_expecteds(triggers, source_file, ts);
  rewrite_expecteds_method(expecteds);

  let transformed = source_text;
  expecteds.slice().reverse().forEach(expected => {
    const assertion_expression =
      expected.flag === AssertionFlag.Show
        ? `console.log(${expected.expression})`
        : expected.flag === AssertionFlag.Fail
          ? `expect(function () { return ${expected.expression}; }).toThrowError()`
          : expected.flag === AssertionFlag.Pass
            ? `expect(${expected.expression}).toEqual(${expected.value})`
            : `expect(function () { return ${expected.expression}; }).not.toThrowError()`;
    const description = get_formatted_description(expected, true);
    transformed = `${transformed.slice(
      0,
      expected.start,
    )}${`${expected.method}(${JSON.stringify(
      description,
    )}, function () { ${assertion_expression}; })`}${transformed.slice(
      expected.end,
    )}`;
  });

  if (jest_config._dts_jest_internal_test_ === true) {
    return transformed;
  }

  return ts.transpile(
    transformed,
    {
      ...tsconfig,
      inlineSourceMap: true,
    },
    source_filename,
  );
};
