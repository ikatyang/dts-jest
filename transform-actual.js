exports.process = require('./lib/index').transform_actual;
