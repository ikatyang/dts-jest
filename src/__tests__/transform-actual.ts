import * as fs from 'fs';
import * as path from 'path';
import * as ts from 'typescript';
import {transform_actual} from '../transform-actual';

const transform_fixture = (relative_path: string, tsconfig: ts.CompilerOptions = {}, debug = true) => {
  const source_filename = path.resolve(__dirname, relative_path);
  const source_text = fs.readFileSync(source_filename, 'utf8');
  const jest_config = {globals: {_dts_jest_: {tsconfig}}};
  return transform_actual(source_text, source_filename, jest_config, debug);
};

it('should transform correctly', () => {
  expect(transform_fixture('../../fixtures/transform-actual/general.ts')).toMatchSnapshot();
});

it('should throw error with unattachable expected-comment', () => {
  expect(() => transform_fixture('../../fixtures/transform-actual/unattachable.ts')).toThrowErrorMatchingSnapshot();
  expect(() => transform_fixture('../../fixtures/transform-actual/unattachable-mix.ts')).toThrowErrorMatchingSnapshot();
});

it('should transform correctly with group', () => {
  expect(transform_fixture('../../fixtures/transform-actual/group.ts')).toMatchSnapshot();
});

it('should transform correctly with group-only', () => {
  expect(transform_fixture('../../fixtures/transform-actual/group-only.ts')).toMatchSnapshot();
});

describe('es6', () => {
  it('should transform correctly', () => {
    expect(transform_fixture('../../fixtures/transform-actual/es6.ts')).toMatchSnapshot();
  });

  it('should transform correctly for ts.transpile', () => {
    const tsconfig = {
      module: ts.ModuleKind.CommonJS,
      target: ts.ScriptTarget.ES5,
    };
    expect(transform_fixture('../../fixtures/transform-actual/es6.ts', tsconfig, false)).toMatchSnapshot();
  });
});
