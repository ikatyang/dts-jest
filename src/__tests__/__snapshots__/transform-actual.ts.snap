// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`es6 should transform correctly 1`] = `
"import * as path from 'path';

// @dts-jest
test(\\"['path/to/abc.ts'].map(x => path.basename(x))\\", function () { expect(function () { return ['path/to/abc.ts'].map(x => path.basename(x)); }).not.toThrowError(); }) //=> ['abc.ts']
"
`;

exports[`es6 should transform correctly for ts.transpile 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
var path = require(\\"path\\");
// @dts-jest
test(\\"['path/to/abc.ts'].map(x => path.basename(x))\\", function () { expect(function () { return ['path/to/abc.ts'].map(function (x) { return path.basename(x); }); }).not.toThrowError(); }); //=> ['abc.ts']
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiZXM2LmpzIiwic291cmNlUm9vdCI6IiIsInNvdXJjZXMiOlsiZXM2LnRzIl0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiI7O0FBQUEsMkJBQTZCO0FBRTdCLFlBQVk7QUFDWixJQUFJLENBQUMsK0NBQStDLEVBQUUsY0FBYyxNQUFNLENBQUMsY0FBYyxNQUFNLENBQUMsQ0FBQyxnQkFBZ0IsQ0FBQyxDQUFDLEdBQUcsQ0FBQyxVQUFBLENBQUMsSUFBSSxPQUFBLElBQUksQ0FBQyxRQUFRLENBQUMsQ0FBQyxDQUFDLEVBQWhCLENBQWdCLENBQUMsQ0FBQyxDQUFDLENBQUMsQ0FBQyxDQUFDLEdBQUcsQ0FBQyxZQUFZLEVBQUUsQ0FBQyxDQUFDLENBQUMsQ0FBQyxDQUFBLENBQUMsZUFBZSJ9"
`;

exports[`should throw error with unattachable expected-comment 1`] = `
"Unattachable expected-value(s) detected:

  fixtures/transform-actual/unattachable.ts:1
  fixtures/transform-actual/unattachable.ts:3
  fixtures/transform-actual/unattachable.ts:5
  fixtures/transform-actual/unattachable.ts:7
  fixtures/transform-actual/unattachable.ts:9"
`;

exports[`should throw error with unattachable expected-comment 2`] = `
"Unattachable expected-value(s) detected:

  fixtures/transform-actual/unattachable-mix.ts:1
  fixtures/transform-actual/unattachable-mix.ts:3
  fixtures/transform-actual/unattachable-mix.ts:7
  fixtures/transform-actual/unattachable-mix.ts:9
  fixtures/transform-actual/unattachable-mix.ts:11
  fixtures/transform-actual/unattachable-mix.ts:13"
`;

exports[`should transform correctly 1`] = `
"// @dts-jest
test.skip(\\"Math.max(1, 2, 3)\\", function () { expect(function () { return Math.max(1, 2, 3); }).not.toThrowError(); }) //=> 3

// @dts-jest description-none
test.skip(\\"description-none\\", function () { expect(function () { return Math.max(1, 2, 3); }).not.toThrowError(); }) //=> 3

// @dts-jest:show
test.skip(\\"Math.min(1, 2, 3)\\", function () { console.log(Math.min(1, 2, 3)); }) //=> 1

// @dts-jest:show description-show
test.skip(\\"description-show\\", function () { console.log(Math.min(1, 2, 3)); }) //=> 1

// @dts-jest:skip
test.skip(\\"Math.abs(-1)\\", function () { expect(function () { return Math.abs(-1); }).not.toThrowError(); }) //=> 1

// @dts-jest:skip description-skip
test.skip(\\"description-skip\\", function () { expect(function () { return Math.abs(-1); }).not.toThrowError(); }) //=> 1

// @dts-jest:only
test(\\"Math.floor(1.2)\\", function () { expect(function () { return Math.floor(1.2); }).not.toThrowError(); }) //=> 1

// @dts-jest:only description-only
test.skip(\\"description-only\\", function () { expect(function () { return Math.floor(1.2); }).not.toThrowError(); }) //=> 1

// @dts-jest:pass
test.skip(\\"Math.abs(-2)\\", function () { expect(Math.abs(-2)).toEqual(2); }) //=> 2

// @dts-jest:pass description-pass
test.skip(\\"description-pass\\", function () { expect(Math.abs(-2)).toEqual(2); }) //=> 2

// @dts-jest:fail
test.skip(\\"Math.round(1.8)\\", function () { expect(function () { return Math.round(1.8); }).toThrowError(); }) //=> 2

// @dts-jest:fail description-fail
test.skip(\\"description-fail\\", function () { expect(function () { return Math.round(1.8); }).toThrowError(); }) //=> 2

// @dts-jest
test.skip(\\"Math.max(\\\\n      1,\\\\n      2,\\\\n      3,\\\\n    )\\", function () { expect(function () { return Math.max(
  1,
  2,
  3,
); }).not.toThrowError(); }) //=> 3
"
`;

exports[`should transform correctly with group 1`] = `
"// @dts-jest:group A

// @dts-jest:show
test(\\"Math.min(1, 2, 3)\\", function () { console.log(Math.min(1, 2, 3)); }) //=> 1

// @dts-jest:show description-show
test(\\"description-show\\", function () { console.log(Math.min(1, 2, 3)); }) //=> 1

// @dts-jest:skip
test.skip(\\"Math.abs(-1)\\", function () { expect(function () { return Math.abs(-1); }).not.toThrowError(); }) //=> 1

// @dts-jest:skip description-skip
test.skip(\\"description-skip\\", function () { expect(function () { return Math.abs(-1); }).not.toThrowError(); }) //=> 1

// @dts-jest:group:skip B

// @dts-jest:only
test.skip(\\"Math.floor(1.2)\\", function () { expect(function () { return Math.floor(1.2); }).not.toThrowError(); }) //=> 1

// @dts-jest:only description-only
test.skip(\\"description-only\\", function () { expect(function () { return Math.floor(1.2); }).not.toThrowError(); }) //=> 1

// @dts-jest:pass
test.skip(\\"Math.abs(-2)\\", function () { expect(Math.abs(-2)).toEqual(2); }) //=> 2

// @dts-jest:group C

// @dts-jest:pass description-pass
test(\\"description-pass\\", function () { expect(Math.abs(-2)).toEqual(2); }) //=> 2

// @dts-jest:fail
test(\\"Math.round(1.8)\\", function () { expect(function () { return Math.round(1.8); }).toThrowError(); }) //=> 2

// @dts-jest:fail description-fail
test(\\"description-fail\\", function () { expect(function () { return Math.round(1.8); }).toThrowError(); }) //=> 2

// @dts-jest
test(\\"Math.max(\\\\n      1,\\\\n      2,\\\\n      3,\\\\n    )\\", function () { expect(function () { return Math.max(
  1,
  2,
  3,
); }).not.toThrowError(); }) //=> 3
"
`;

exports[`should transform correctly with group-only 1`] = `
"// @dts-jest
test.skip(\\"Math.max(1, 2, 3)\\", function () { expect(function () { return Math.max(1, 2, 3); }).not.toThrowError(); }) //=> 3

// @dts-jest description-none
test.skip(\\"description-none\\", function () { expect(function () { return Math.max(1, 2, 3); }).not.toThrowError(); }) //=> 3

// @dts-jest:group A

// @dts-jest:show
test.skip(\\"Math.min(1, 2, 3)\\", function () { console.log(Math.min(1, 2, 3)); }) //=> 1

// @dts-jest:show description-show
test.skip(\\"description-show\\", function () { console.log(Math.min(1, 2, 3)); }) //=> 1

// @dts-jest:skip
test.skip(\\"Math.abs(-1)\\", function () { expect(function () { return Math.abs(-1); }).not.toThrowError(); }) //=> 1

// @dts-jest:only description-skip
test.skip(\\"description-skip\\", function () { expect(function () { return Math.abs(-1); }).not.toThrowError(); }) //=> 1

// @dts-jest:group:only B

// @dts-jest:only
test(\\"Math.floor(1.2)\\", function () { expect(function () { return Math.floor(1.2); }).not.toThrowError(); }) //=> 1

// @dts-jest:only description-only
test.skip(\\"description-only\\", function () { expect(function () { return Math.floor(1.2); }).not.toThrowError(); }) //=> 1

// @dts-jest:pass
test.skip(\\"Math.abs(-2)\\", function () { expect(Math.abs(-2)).toEqual(2); }) //=> 2

// @dts-jest:group C

// @dts-jest:pass description-pass
test.skip(\\"description-pass\\", function () { expect(Math.abs(-2)).toEqual(2); }) //=> 2

// @dts-jest:fail
test.skip(\\"Math.round(1.8)\\", function () { expect(function () { return Math.round(1.8); }).toThrowError(); }) //=> 2

// @dts-jest:fail description-fail
test.skip(\\"description-fail\\", function () { expect(function () { return Math.round(1.8); }).toThrowError(); }) //=> 2

// @dts-jest
test.skip(\\"Math.max(\\\\n      1,\\\\n      2,\\\\n      3,\\\\n    )\\", function () { expect(function () { return Math.max(
  1,
  2,
  3,
); }).not.toThrowError(); }) //=> 3
"
`;
